### DISCUSSION TITLE
Solved: netbox-rq fails to start after upgrade from 3.6.1 to 3.7.3 (cannot import name 'Connection' from 'rq')

### DISCUSSION THREAD

**Discussion by yngmike** [netbox-community/netbox#18171] [Help Wanted!] [answered]
EDIT:

Solved:
Adding: "rq==1.16.2"
To the top of /opt/netbox/requirements.txt and re-running upgrade.sh fixed the problem.

Hello, I'm trying to align my devs lab environments to match our prod installation which is 3.7.3
I am far from a web developer, just a lowly Linux Admin.
It seems that there was a very similar error reported when rq deprecated a DB connection call in the past.
I presume this the new method, but it still seems to fail.
upgrade.sh completes without errors and the main netbox process starts and runs.

Any help would be appreciated.
Thanks,
-Mike

OS: RHEL 8.8
Django-rq

Requirement already satisfied: django-rq in /usr/local/lib/python3.9/site-packages (3.0.0)
Requirement already satisfied: rq>=2 in /usr/local/lib/python3.9/site-packages (from django-rq) (2.0.0)
Requirement already satisfied: django>=3.2 in /usr/local/lib/python3.9/site-packages (from django-rq) (4.2.5)
Requirement already satisfied: redis>=3.5 in /usr/local/lib/python3.9/site-packages (from django-rq) (5.0.0)
Requirement already satisfied: click>=5 in /usr/local/lib/python3.9/site-packages (from rq>=2->django-rq) (8.1.7)
Requirement already satisfied: sqlparse>=0.3.1 in /usr/local/lib/python3.9/site-packages (from django>=3.2->django-rq) (0.4.4)
Requirement already satisfied: asgiref<4,>=3.6.0 in /usr/local/lib/python3.9/site-packages (from django>=3.2->django-rq) (3.7.2)
Requirement already satisfied: async-timeout>=4.0.2; python_full_version <= "3.11.2" in /usr/local/lib/python3.9/site-packages (from redis>=3.5->django-rq) (4.0.3)
Requirement already satisfied: typing-extensions>=4; python_version < "3.11" in /usr/local/lib/python3.9/site-packages (from asgiref<4,>=3.6.0->django>=3.2->django-rq) (4.8.0)

Error:

Dec  6 12:59:32 cn04tlv-netboxcbad01 systemd[1]: Started NetBox Request Queue Worker.
Dec  6 12:59:33 cn04tlv-netboxcbad01 python3[12686]: Traceback (most recent call last):
Dec  6 12:59:33 cn04tlv-netboxcbad01 python3[12686]:  File "/opt/netbox/netbox/manage.py", line 10, in <module>
Dec  6 12:59:33 cn04tlv-netboxcbad01 python3[12686]:    execute_from_command_line(sys.argv)
Dec  6 12:59:33 cn04tlv-netboxcbad01 python3[12686]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/core/management/__init__.py", line 442, in execute_from_command_line
Dec  6 12:59:33 cn04tlv-netboxcbad01 python3[12686]:    utility.execute()
Dec  6 12:59:33 cn04tlv-netboxcbad01 python3[12686]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/core/management/__init__.py", line 436, in execute
Dec  6 12:59:33 cn04tlv-netboxcbad01 python3[12686]:    self.fetch_command(subcommand).run_from_argv(self.argv)
Dec  6 12:59:33 cn04tlv-netboxcbad01 python3[12686]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/core/management/__init__.py", line 275, in fetch_command
Dec  6 12:59:33 cn04tlv-netboxcbad01 python3[12686]:    klass = load_command_class(app_name, subcommand)
Dec  6 12:59:33 cn04tlv-netboxcbad01 python3[12686]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/core/management/__init__.py", line 48, in load_command_class
Dec  6 12:59:33 cn04tlv-netboxcbad01 python3[12686]:    module = import_module("%s.management.commands.%s" % (app_name, name))
Dec  6 12:59:33 cn04tlv-netboxcbad01 python3[12686]:  File "/usr/lib64/python3.9/importlib/__init__.py", line 127, in import_module
Dec  6 12:59:33 cn04tlv-netboxcbad01 python3[12686]:    return _bootstrap._gcd_import(name[level:], package, level)
Dec  6 12:59:33 cn04tlv-netboxcbad01 python3[12686]:  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
Dec  6 12:59:33 cn04tlv-netboxcbad01 python3[12686]:  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
Dec  6 12:59:33 cn04tlv-netboxcbad01 python3[12686]:  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
Dec  6 12:59:33 cn04tlv-netboxcbad01 python3[12686]:  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
Dec  6 12:59:33 cn04tlv-netboxcbad01 python3[12686]:  File "<frozen importlib._bootstrap_external>", line 850, in exec_module
Dec  6 12:59:33 cn04tlv-netboxcbad01 python3[12686]:  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
Dec  6 12:59:33 cn04tlv-netboxcbad01 python3[12686]:  File "/opt/netbox/netbox/core/management/commands/rqworker.py", line 3, in <module>
Dec  6 12:59:33 cn04tlv-netboxcbad01 python3[12686]:    from django_rq.management.commands.rqworker import Command as _Command
Dec  6 12:59:33 cn04tlv-netboxcbad01 python3[12686]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django_rq/management/commands/rqworker.py", line 5, in <module>
Dec  6 12:59:33 cn04tlv-netboxcbad01 python3[12686]:    from rq import Connection
Dec  6 12:59:33 cn04tlv-netboxcbad01 python3[12686]: **ImportError: cannot import name 'Connection' from 'rq' (/opt/netbox/venv/lib64/python3.9/site-packages/rq/__init__.py)**

If I attempt to downgrade RQ I get multiple dependency errors.

**django-rq 3.0.0 requires rq>=2, but you'll have rq 1.16.2 which is incompatible.**

**Comment by candlerb** [comment, marked as answer]
> Requirement already satisfied: django-rq in /usr/local/lib/python3.9/site-packages (3.0.0)

You don't appear to be installing Netbox in a virtualenv. If not, it is bound to end in tears, as the versions of python packages which are installed globally on your system are almost certainly not compatible with those Netbox requires.

You should use the script `/opt/netbox/upgrade.sh` to install the packages; this will wipe and recreate a fresh virtualenv for you, and it *should* ignore any globally-installed versions of packages.

Then, if you *still* can't get the upgrade to work, there might be other problems. Unfortunately, Netbox doesn't pin its dependencies, so the packages you get when you install Netbox 3.7.3 today are not the same as the packages you would have got when 3.7.3 was first released - and the newer packages are known to break things.

You can find a working set of dependencies from here: https://github.com/markkuleinio/netbox-freezes - these are taken from points in time when those versions of Netbox *could* be installed successfully. I can see that version 3.7.3 worked with:

```
django-rq==2.10.1
rq==1.15.1
```

but in v3.7.3, requirements.txt pinned django-rq==2.10.1 but did not pin rq.  It might be sufficient just to edit requirements.txt and add the dependency on that version of rq.

A similar change was added recently to Netbox master in commit f568433, dated Oct 28 2024, pinning rq==1.16.2, which made it into Netbox v4.1.5. However this was not back-ported to any earlier versions of Netbox (indeed, I've never seen a retrospective update to any version of Netbox other than the current master version)

So basically, if you're using anything other than the most recently-released version of Netbox, you're expected to work out all this dependency stuff yourself. But of course, to get yourself onto the most recent version, you still have to upgrade to v3.7.x first. I think there would be a case for a v3.7.9 release to address this particular problem, or if not, then a documented workaround.

(Aside: you might as well go straight to v3.7.8, rather than v3.7.3, to get the latest set of bug fixes on that particular series)

**Reply by yngmike** [reply]
candlerb, Thank you so much for the quick reply.

pinning "rq==1.16.2" at the top or requirements.txt fixed all of the problems (no errors in /var/log/messages).
upgrade.sh runs successfully
I have passed it off to my dev team. Thank you for pointing out that global packages have no effect on the virtualenv, which I am using.
upgrade.sh runs successfully 

Have a great day!
